### ISSUE TITLE
Source generators slow down IntelliSense session activation

### ISSUE THREAD

**Issue by kindermannhubert** [dotnet/roslyn#53889]
**Version Used**:  VS 16.10.0

For a long time, I've been observing slow reaction times of IntelliSense in our code base. For example when you're adding a new field to a class and want to write
```private readonly```
it should be (and normally is) sufficient to just write the following sequence of characters
```'p' + 'r' + 'i' + space + 'r' + 'e' + space``` 
(or something similar; probably depends on your settings and last used items). Both words should be completed with typing of space char.
But when IntelliSense is slowed down it will not manage to activate before pressing the space and the result will be exactly what you have written:
```pri re ```
and it's annoying as hell. When you want 'correct' behavior you have to slow down your typing😞. 

I've spent a day examining why it's sometimes so slow. I tried disabling extensions, analyzers, tweaking settings, but nothing worked. Finally, I tracked it down to source generators. I believe when a character is typed, source generators (at least on all project files) execution precedes IntelliSense activation. 
I've tried adding Thread.Sleep to the generator and it postponed activation of IntelliSense after character typing. Moreover, direct triggering of IntelliSense for example by Ctrl+J is fast (because it's not triggering SG). 

The slowdown is not anything extreme - I believe it's under 1s. But if you write "fast enough" you're in trouble.
There doesn't have to be anything crazy happening in the generator. I was trying (just to test) to shave it off so it was doing almost nothing (certainly not generating anything, just checking if types have some attribute and getting some type symbols from the compilation). 

The project I was testing it in has a little over 2k C# files.

**Expected Behavior**:
IntelliSense activation time is not affected by used source generators.

**Actual Behavior**:
IntelliSense activation time is affected by used source generators.

**Comment by Youssef1313** [comment]
The proposal in https://github.com/dotnet/roslyn/issues/51497 may fix that.

---

> The slowdown is not anything extreme - I believe it's under 1s. But if you write "fast enough" you're in trouble.

In case you really mean "one second", then this is **very** extreme. See:

![image](https://user-images.githubusercontent.com/31348972/120884602-470a3a00-c5e4-11eb-872a-0c52f72ce9bf.png)

**Comment by kindermannhubert** [comment]
Yeah, I've meant one second 🙂 (completion session activation delay). Problem is that even if SG is "fast" its execution time depends on project size. So even the fastest SGs become slow for large code bases. #53452 describes this problem well.

**Comment by sharwell** [comment]
> But when IntelliSense is slowed down it will not manage to activate before pressing the space and the result will be exactly what you have written:

Yes, I also find this "feature" to be disruptive. You can turn it off here:
![image](https://user-images.githubusercontent.com/1408396/122304882-cb996a00-ceba-11eb-83f6-a361d0149124.png)

If you experience typing delays as a result of this change, you can file feedback using these steps:
https://docs.microsoft.com/en-us/visualstudio/ide/how-to-increase-chances-of-performance-issue-being-fixed?view=vs-2019#slowness-and-high-cpu-issues

**Comment by kindermannhubert** [comment]
Thanks, but the suggested solution does not help.
Not sure if you suggest turning off the two settings or enable but I tried different combinations.

"Responsive code completion" - haven't found any documentation, but looks like it enables for example completion after pressing space. So turning it off is unacceptable for me (but turning it off does not help either).

Turning on "Auto cancel ..." does not change anything on the described problem. I tried both - auto and manual settings.

Currently, we solve this by a super ugly hack when we check the current call stack in ISourceGenerator.Initialize and try to find out if it's called from IntelliSense or not. If it's called from IS we don't register for any actions (because RegisterForSyntaxNotifications is sufficient for described slow down). This is possible only because this SG generates some explicit interface implementations (so it does not matter if it's missing from IS compilation).
This particular hack could be done much more nicely if MsBuild properties would be available in ISourceGenerator.Initialize but it's not. They are unfortunately available only in ISourceGenerator.Execute.

There is simply too much work happening after typing a letter. You just can't visit every syntax node of a large project and expect it will keep up with faster typing. It would be needed to get more fine-grained control over what SG is going to analyze and when (but that's already mentioned in other issues).

**Comment by CyrusNajmabadi** [comment]
> There is simply too much work happening after typing a letter. You just can't visit every syntax node of a large project and expect it will keep up with faster typing. It would be needed to get more fine-grained control over what SG is going to analyze and when (but that's already mentioned in other issues).

This is correct.   We tried to make it clear early on that expensive source generators were not a fit for the SG1.0 api.  Indeed, performance of tooling was one of the top issues preventing us from even pursuing this space until recently.  We only did 1.0 to have a mechanism for cheap generators to be viable.  The incremental generators approach is what we will be recommending for expensive analyzers that need to do minimal incremental work, and may have parts of them that should not run in scenarios taht do not need the information they would generate.
